### ISSUE TITLE
withMulti results become undefined on terms change

### ISSUE THREAD

**Issue by eric-burel** [VulcanJS/Vulcan#2090]
Hi, all is in the title. I setup a Map component that rely on withMulti. The architecture is:
- The parent component wraps its children with `withMulti` (as the Datatable does with DatatableContent)
- The children display results on map
- There is an intermediate component that manage `terms` update when the user moves

Terms are correctly updated. But if I move the Map, the marker disappear, as results become `undefined` (ie `graphql` `props.data[resolverName]` is undefined). No query is triggered according to network. A call to `refetch` fixes but provokes blinking and sounds like a palliative.

However, when the next polling is triggered, the correct data are retrieved. It's just changing `terms` that does not behave as expected.

This issue seems to have appeared to other people, but not all the time, so it seems related to how we define the React component and how we use `withMulti`.
I can reproduce even with `apollo-client` and `apollo-react` version 2 as well as the current Vulcan version. Maybe smth is failing silently ?

I will try to setup a minimal repro, but in the meantime here is my component code, in case someone  understand this issue (or want to setup a Leaflet Map that almost work :)):

```jsx
import React, { Component } from 'react';
import { Components, registerComponent, withMulti } from 'meteor/vulcan:core';

// the parent, it creates the component wrapped with `withMulti` and 
// handle the initial render
class GeoSearchPage extends Component {
  render() {
    const { collection, defaultTerms, ...otherProps } = this.props;
    const withMultiOptions = {
      collection,
      terms: defaultTerms
    };
    const GeoSearchPageInnerWithMulti = withMulti(withMultiOptions)(
      Components.GeoSearchPageInner
    );
    return (
      <ViewportProvider
        {...otherProps}
        defaultTerms={defaultTerms}
        C={GeoSearchPageInnerWithMulti}
      />
    );
  }
}

// manage state, is in charge of updating terms and passing to its child,
// which is a component wrapped with withMulti
class ViewportProvider extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentTerms: props.defaultTerms,
      currentViewport: null
    };
  }
  isValidViewport = viewport => {
    return (
      typeof viewport.center[0] === 'number' &&
      typeof viewport.center[1] === 'number' &&
      viewport.zoom < Infinity
    );
  };
  shouldUpdateTerms = newViewport => {
    const { currentViewport } = this.state;
    if (!this.isValidViewport(newViewport)) return false;
    if (!currentViewport && newViewport) return true;
    return (
      newViewport.center[0] !== currentViewport.center[0] ||
      newViewport.center[1] !== currentViewport.center[1] ||
      newViewport.zoom !== currentViewport.zoom
    );
  };
  computeNewTerms = viewport => {
    const { zoom, center } = viewport;
    const [lat, lng] = center;
    const terms = { ...this.state.currentTerms };
    terms.geoQuery.point.lat = lat;
    terms.geoQuery.point.lng = lng;
    terms.geoQuery.maxDistance = 10000; // TODO compute depending on zoom
    return terms;
  };
  onViewportChanged = viewport => {
    const newTerms =
      this.shouldUpdateTerms(viewport) && this.computeNewTerms(viewport);
    const newViewport = this.isValidViewport(viewport) && viewport;
    let newState = {};
    if (newTerms) newState['currentTerms'] = newTerms;
    if (newViewport) newState['currentViewport'] = newViewport;
    this.setState(newState);
  };

  render() {
    const { C, ...otherProps } = this.props;
    const { currentViewport, currentTerms: terms } = this.state;
    const additionalProps = {
      viewport: currentViewport,
      terms: terms,
      onViewportChanged: this.onViewportChanged
    };
  // C is a component wrapped with `withMulti`, we pass the terms props to it
    return <C {...otherProps} {...additionalProps} />;
  }
}

// some helper, ignore it
const computePositions = (results, latLngField = 'placeLatLng') => {
  return (
    results.filter(r => !!r[latLngField]).map(r => r[latLngField])
      .filter(
        ({ lat, lng }) => typeof lat === 'number' && typeof lng === 'number'
      ).map(({ lat, lng }) => [lat, lng])
  );
};

// inner map, that will be wrapped with withMulti
class GeoSearchPageInner extends Component {
  componentWillReceiveProps(newProps) {
    if (newProps.terms !== this.props.terms) {
       // will work, but the marker still blink as `results` is temporarily undefined
      // refetch should not be necessary I think
      this.props.refetch()
    }
  }
  render() {
    const {
      results = [],
      onViewportChanged,
      ...otherProps
    } = this.props;
    const positions = computePositions(results);
    return (
      <div>
        <Components.LeafletMap
          positions={positions}
          onViewportChanged={onViewportChanged}
          {...otherProps}
        />
      </div>
    );
  }
}

registerComponent({
  name: 'GeoSearchPageInner',
  component: GeoSearchPageInner
});
registerComponent({ name: 'GeoSearchPage', component: GeoSearchPage });
```

**Comment by eric-burel** [comment]
Hi @meteorplus, no it's not related to polling (even disabling it does not change anything). Changes in the `terms` prop should trigger a new query automatically, whatever the polling interval is, which is not the case here. Then the polling is triggered and the data reappear, which shows that the query itself is ok.

I suspect that it comes from how `react-apollo` `graphql` tag handles changes in `variables`, it may not be directly related to Vulcan, but I prefer to open an issue because it might be solved by tweaking the `withMulti` HOC.

**Comment by eric-burel** [comment]
It's a custom view already, that transforms the params into a geoquery (not in the code here). I'll try to change it when I come back at this issue, since it is currently overly complex anyway.

**Comment by Apollinaire** [comment]
Update on this : I've noticed that it was happening consistently with the datatable sort when you set or change `orderBy` in the `withMulti` terms.

The strange thing is that when you set `orderBy: { firstName: 1 }` the props at withMulti become `{ firstName: 1, _id: 1 }` then I found out that this was done in `collection.prototype.getParameters` to break ties between elements. Fine. But it seems that this is what is breaking stuff. **We are changing the props in the middle of the query and it seems to bother apollo client**.

So here is what I did : instead of setting `orderBy : { firstName : 1 }` I add the `_id: -1` myself everytime I want to sort something. So when the hoc updates it does not add it itself.
This is just a workaround, we still have to find a way to avoid this behaviour for every case.
The good (or hard) thing to do here is that we have to find a way to set the right terms on the client (because of apollo cache) without messing the components' logic.

I'll submit a PR to patch the datatable's sort.

TLDR: If you encounter this bug, look at the terms on the withMulti component, and see if they differ from the terms you pass to it. Set the terms you pass it to match those on withMulti.

**Comment by eric-burel** [comment]
@Apollinaire would writing a `withTerms` hoc help here? 

My idea is to add a layer of abstraction, so that the `terms` passed by the component are never directly passed to the `withMulti`, but instead systematically cleaned up by the `withTerms` component. You could pass it default terms, the same way we currently do with `withMulti`, and it would provide terms update methods. 
If it needs to add an `id` for whatever purpose, then it could do it there, so that the query is always consistent no matter how the component handle terms. Does it sounds doable to you?

I already proposed this to simplify list components logic anyway (DataTable, custom components etc.), because they basically all handle terms the same way and share the same needs (search, sort, and handle custom terms).
It could even be merged with `withMulti`, so that we get a single hoc to manage both data fetching and the terms configuration.

**Comment by Apollinaire** [comment]
Doing this in an hoc might work indeed, because this specific task is not dependent on the result of the query, so it does not have to happen in the reducer of the `graphql` hoc, like it is now.
I think it's something for recompose's `withProps` ? https://github.com/acdlite/recompose/blob/master/docs/API.md#withprops
I'm on it

**Comment by SachaG** [comment]
We should probably test this on the `apollo2` branch? Who knows, maybe Apollo Client 2.0 doesn't have this issue?

**Comment by Apollinaire** [comment]
i think it can be done here, it should not break things: the only thing we are doing is taking some work out of the queryReducer and doing it above, the same functions will be ran with the same arguments, just a bit earlier in the component tree. I'm doing the PR, this way we can test it and see if we keep it or not

**Comment by Apollinaire** [comment]
Update: I tried to pre-compute the terms out of the query reducer : this is just above the `graphql()` hoc in `compose`, in withMulti :
```js
    mapProps( ({terms, client, paginationTerms, ...rest}) => {
      const mergedTerms = { ...options.terms, ...terms, ...paginationTerms}
     const result = collection.getParameters(mergedTerms, client);
     return {terms: result, client, paginationTerms, ...rest}
    
    } ),
```
And I removed this logic out of the query reducer. This way we should have the right terms from the beginning and we should not have to recompute them each time. It does not work, and breaks the sort. 
I don't have more time to give to this, so I will instead just fix the datatable's sort feature and maybe we will try to fix it after the apollo 2 update. [The `reducer` option config seems deprecated](https://www.apollographql.com/docs/react/recipes/2.0-migration.html#reducers) so we will have to rewrite the `withMulti` hoc

**Comment by SachaG** [comment]
>  The reducer option config seems deprecated so we will have to rewrite the withMulti hoc

Yes, that's already done on the Apollo 2 branch.
